src/units/abilities.cpp

```diff
@@ -1557,30 +1557,54 @@ bool attack_type::special_active_impl(const_attack_ptr self_attack, const_attack
 	temporary_facing other_facing(other, other_loc.get_relative_dir(self_loc));
 
 	// Filter poison, plague, drain, first strike
-	if (tag_name == "drains" && other && other->get_state("undrainable")) {
+	if (tag_name == "drains" && special_affects_self(special, is_attacker) && other && other->get_state("undrainable")) {
 		return false;
 	}
-	if (tag_name == "plague" && other &&
+	if (tag_name == "drains" && special_affects_opponent(special, is_attacker) && self && self->get_state("undrainable")) {
```

Comment:
Should this depend on `special_active_impl`'s `whom` argument to work out which logic to check?

As the `(affects_self && other && other->get_state)` and `affects_opponent && self && self->get_state` logic is repeated many times, I'm thinking it would be better to create a local lambda function so that this `if` could become
```
if(tag_name == "drains" && lambda_name("undrainable"))
```
and later ones could become
```
if(tag_name == "poison" && lambda_name("unpoisonable", unit::STATE_POISONED))
```

Reply:
changes done

Reply:
I meant a [lambda function](https://en.cppreference.com/w/cpp/language/lambda). However, the new code is simpler than a lambda would be, so there's no need to make more changes for this.